### PR TITLE
feat(upgrade): Improve upgrade framework

### LINKED ISSUE #1219: Add version property to knora-base and check on load ontologies

### LINKED ISSUE #1336: Upgrade scripts
The upgrade scripts are named after PRs and by what they are doing.

Both informations don't tell me which one I need to run when I'm running version `x.y.z`.

For each release, it would be nice to have a list of upgrade scripts that need to be 

Alternatively, I would very much prefer, if it would be possible to only have one script that needs to be run. This script could then figure out by itself based on a version string in knora-base which of those upgrades need to be run and then do it.

I would also need to be able to package it inside a docker container and run it this way on the server.

### PR DESCRIPTION
**Pull request by benjamingeer**
This PR introduces a single script, `upgrade/update-repository.py`, for updating repository data when Knora is upgraded. The script checks which updates are necessary with the current version of Knora, and applies only those updates.

All updates are done offline rather than in SPARQL, because this is often faster, allows the script to log its progress, and avoids potential differences in SPARQL behaviour between triplestore versions.

Each PR-specific update is implemented as a Python plug-in class that is dynamically loaded by the main script as needed. Each plugin is in a subdirectory of `upgrade/plugins`, whose name contains the PR number. The main script and the plugins use library functions in `upgrade/updatelib/rdftools.py`.

This PR repackages the updates for PRs merged since release 7.0.0, turning them into plugins in the new framework:

- #1307 
- #1322 

The plugins have their own unit tests. To run them, run `./test.sh` in the `upgrade` directory (uses `pytest`).

The `knora-base` ontology has a version string that allows the update scripts to check which updates are needed.

When Knora starts, `OntologyResponderV2` checks that `knora-base` has the expected version string.

User documentation is in `docs/src/paradox/04-deployment/updates.md`.

Resolves #1336.
Resolves #1219.

### CODE REVIEW

**Comment by benjamingeer** [comment]
TODO:

- Try to preserve prefixes in temporary files.
- Decide what to do about testing.

File: upgrade/update-repository.py [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+    return transformers
+
+
+# Command-line invocation.
+def main():
+    default_graphdb_host = "localhost"
+    default_repository = "knora-test"
+
+    parser = argparse.ArgumentParser(description="Updates a Knora repository.")
+    parser.add_argument("-g", "--graphdb", help="GraphDB host (default '{}')".format(default_graphdb_host), type=str)
```
Comment (subotic) [inline upgrade/update-repository.py]: You can add default values with the `default` parameter. Here is an example:

https://github.com/dasch-swiss/0807-mls-scripts/blob/7e3623b623a347ea254f90807dec17c5a34a7d17/mls-import.py#L559

Comment (benjamingeer) [reply upgrade/update-repository.py]: Thanks, done in e39736c.

**Comment by benjamingeer** [comment]
> Could you maybe add a `requirements.txt`

Good idea, thanks, done in e39736c.

**Comment by benjamingeer** [comment]
@tobiasschweizer Would you like to try this new repository update script? I got it to preserve Turtle prefixes.

**Comment by benjamingeer** [comment]
@subotic I would like Travis to run `./test.sh` in the `upgrade` directory when it runs the other Knora tests, but can't figure out where to put that in `.travis.yml`. What would you recommend?

**Comment by subotic** [comment]
> @subotic I would like Travis to run `./test.sh` in the `upgrade` directory when it runs the other Knora tests, but can't figure out where to put that in `.travis.yml`. What would you recommend?

So only GraphDB would need to run? I would say, create a new stage:

```
    # upgrade tests
    - stage: test
      script:
        # prepare needed graphdb-se files
        - mkdir -p $TRAVIS_BUILD_DIR/graphdb
        - cp $TRAVIS_BUILD_DIR/travis/graphdb.license $TRAVIS_BUILD_DIR/graphdb/graphdb.license
        - cp $TRAVIS_BUILD_DIR/webapi/scripts/KnoraRules.pie $TRAVIS_BUILD_DIR/graphdb
        # start and initialize graphdb-se
        - docker run -d -p 127.0.0.1:7200:7200 -v $TRAVIS_BUILD_DIR/graphdb:/graphdb -e GDB_HEAP_SIZE=$GDB_HEAP_SIZE ontotext/graphdb:$GRAPHDB_VERSION -Dgraphdb.license.file=/graphdb/graphdb.license
        - sleep 15
        // run tests
        - cd $TRAVIS_BUILD_DIR/upgrade && ./test.sh
```

You can copy and paste it under the `integration tests` stage.

**Comment by benjamingeer** [comment]
> So only GraphDB would need to run? 

The tests don't need GraphDB, they're self-contained.

**Comment by subotic** [comment]
Then this should suffice:

```
    # upgrade tests
    - stage: test
      script:
        // run tests
        - cd $TRAVIS_BUILD_DIR/upgrade && ./test.sh
```

**Comment by benjamingeer** [comment]
@subotic Thanks, let's see if it works.

I also added a check when Knora starts to make sure that `knora-base` has the expected version number (#1219).

**Comment by subotic** [comment]
> I also added a check when Knora starts to make sure that `knora-base` has the expected version number (#1219).

cool :-)

**Comment by tobiasschweizer** [comment]
> @tobiasschweizer Would you like to try this new repository update script? I got it to preserve Turtle prefixes.

I would actually like too, but unfortunately I am very busy finishing up the BEOL project ...

**Comment by benjamingeer** [comment]
@subotic Strange, it didn't work on Travis:

```
$ {:script=>["cd $TRAVIS_BUILD_DIR/upgrade && ./test.sh"], :stage=>"test"}
/home/travis/.travis/functions: line 104: [cd /home/travis/build/dhlab-basel/Knora/upgrade && ./test.sh],: No such file or directory
The command "{:script=>["cd $TRAVIS_BUILD_DIR/upgrade && ./test.sh"], :stage=>"test"}" exited with 1.
```

https://travis-ci.org/dhlab-basel/Knora/jobs/544657607

**Comment by benjamingeer** [comment]
@subotic How do I install Python libraries on Travis?

```
$ cd $TRAVIS_BUILD_DIR/upgrade && ./test.sh
/usr/bin/python3: No module named pytest
The command "cd $TRAVIS_BUILD_DIR/upgrade && ./test.sh" exited with 1.
```

**Comment by benjamingeer** [comment]
> The command "cd $TRAVIS_BUILD_DIR/upgrade && pip3 install -r requirements.txt && pytest -W ignore::DeprecationWarning" exited with 2.

I think the problem here is that `pytest` doesn't include the current directory in the Python module search path. That's why I made the script `test.sh`, which uses `python3 pytest`. When you run it that way, the current directory is included in the search path, so the modules can be found.

**Comment by subotic** [comment]
I've changed Travis to run `test.sh`. It will hopefully work now.

**Comment by benjamingeer** [comment]
It worked, thanks! Is this OK to merge now?

**Comment by subotic** [comment]
Could you maybe add somewhere in the docs what the system admin needs to run for which version of Knora?

We have the following release versions: `7.0.0`, `6.0.1`, `6.0.0`, `5.0.0`, `4.0.0`, `3.0.0`. Only those for which we have upgrade scripts. This should alleviate some pain and uncertainty from the upgrade process.

Then it should be fine.

**Comment by benjamingeer** [comment]
> Could you maybe add somewhere in the docs what the system admin needs to run for which version of Knora?

Can't we assume that everyone has 7.0.0 already? If so, they just need to read `docs/src/paradox/04-deployment/updates.md `.

**Comment by subotic** [comment]
We could, but we are running `6.0.1` in production, so I would, at least, need exact instructions on what to run on `6.0.1` and in what order. I'm not sure what Lausanne is running at the moment.

**Comment by tobiasschweizer** [comment]
@benjamingeer @subotic I have already converted my data (standoff and value UUID PRs). I assume I need Knora release 8 in production?

**Comment by subotic** [comment]
@tobiasschweizer yes. I would like to release`v8.0.0` in the next few days and then soon afterward upgrade the production server. We will need to coordinate with the `Knora-UI` team.

**Comment by benjamingeer** [comment]
OK, in 442dc14 I documented the steps for upgrades from versions starting at 5.0.0. Before 6.0.0, there were no update scripts, you had to update your data by hand. :)

**Comment by subotic** [comment]
Just running the upgrade script on the data from production. Let's see what happens and how long it takes.

Could please add a parameter so that the `tmp` folder could be supplied?

**Comment by benjamingeer** [comment]
> Could please add a parameter so that the `tmp` folder could be supplied?

According to the [Python docs](https://docs.python.org/3/library/tempfile.html#tempfile.mkstemp), "the user of the application can control the directory location by setting the TMPDIR, TEMP or TMP environment variables", do you want to try that?

the user of the application can control the directory location by setting the TMPDIR, TEMP or TMP environment variables

**Comment by subotic** [comment]
It took 43 without the upload. GraphDB complains since I only ran the main upgrade script. Looks very good. Just need to know what else I need to run.

**Comment by benjamingeer** [comment]
> GraphDB complains since I only ran the main upgrade script.

What error did you get?

> Just need to know what else I need to run.

There's nothing else to run, the main update script should do everything.

**Comment by benjamingeer** [comment]
> without the upload

Do you mean that you typed Control-C before it started doing the upload?

**Comment by benjamingeer** [comment]
If you're running 6.0.1 in production, you have to follow the instructions for that version: 442dc14.

**Comment by subotic** [comment]
> If you're running 6.0.1 in production, you have to follow the instructions for that version

Yes, thanks. Didn't see your post about the instructions. Sometimes the Github page doesn't update. I will try it out as soon as I get back home.

Thank you very much for adding the instructions.

**Comment by subotic** [comment]
> > Could please add a parameter so that the `tmp` folder could be supplied?
> 
> According to the [Python docs](https://docs.python.org/3/library/tempfile.html#tempfile.mkstemp), "the user of the application can control the directory location by setting the TMPDIR, TEMP or TMP environment variables", do you want to try that?

Hmm, yes, better then nothing, but not very precise. Setting the environment variable could have potential side-effects for other tools running. I would prefer an optional command line parameter.

**Comment by subotic** [comment]
The upgrade process has finished. I've started `webapi` and made a few API calls. Looks very good.

@tobiasschweizer I would like to deploy the converted data from production onto our test server. Which BEOL client version do I need to deploy?

**Comment by loicjaouen** [comment]
> I'm not sure what Lausanne is running at the moment.
@subotic : sorry for the lack of reactivity, we are in `v6.0.0`

**Comment by subotic** [comment]
@loicjaouen Ah great. So then you should also be able to upgrade to `8.0.0` after we try it out on our production data.

**Comment by benjamingeer** [comment]
@subotic 

> I would prefer an optional command line parameter.

OK, adding that now.

**Comment by subotic** [comment]
> @subotic
> 
> > I would prefer an optional command line parameter.
> 
> OK, adding that now.

Great thanks. After that, I think we can merge.

**Comment by benjamingeer** [comment]
@subotic Thanks for the review!

**Comment by benjamingeer** [comment]
This PR introduces a single script, `upgrade/update-repository.py`, for updating repository data when Knora is upgraded. The script checks which updates are necessary with the current version of Knora, and applies only those updates.

**Pull request merged**